pkg/cnb/remote_stack_reader.go

```diff
@@ -0,0 +1,154 @@
+package cnb
+
+import (
+	"github.com/google/go-containerregistry/pkg/authn"
+	ggcrv1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/pivotal/kpack/pkg/apis/experimental/v1alpha1"
+	"github.com/pivotal/kpack/pkg/registry/imagehelpers"
+	"github.com/pkg/errors"
+	"strconv"
+	"strings"
+)
+
+const (
+	MixinsLabel = "io.buildpacks.stack.mixins"
+	StackLabel  = "io.buildpacks.stack.id"
+
+	cnbUserId  = "CNB_USER_ID"
+	cnbGroupId = "CNB_GROUP_ID"
+)
+
+type RemoteStackReader struct {
+	RegistryClient RegistryClient
+	Keychain       authn.Keychain
+}
+
+func (r *RemoteStackReader) Read(stackSpec v1alpha1.StackSpec) (v1alpha1.ResolvedStack, error) {
+	buildImage, buildIdentifier, err := r.RegistryClient.Fetch(r.Keychain, stackSpec.BuildImage.Image)
+	if err != nil {
+		return v1alpha1.ResolvedStack{}, err
+	}
+
+	runImage, runIdentifier, err := r.RegistryClient.Fetch(r.Keychain, stackSpec.RunImage.Image)
+	if err != nil {
+		return v1alpha1.ResolvedStack{}, err
+	}
+
+	err = validateStackId(stackSpec.Id, buildImage, runImage)
+	if err != nil {
+		return v1alpha1.ResolvedStack{}, err
+	}
+
+	userId, err := parseCNBID(buildImage, cnbUserId)
+	if err != nil {
+		return v1alpha1.ResolvedStack{}, errors.Wrap(err, "validating build image")
+	}
+
+	groupId, err := parseCNBID(buildImage, cnbGroupId)
+	if err != nil {
+		return v1alpha1.ResolvedStack{}, errors.Wrap(err, "validating build image")
+	}
+
+	buildMixins, err := readMixins(buildImage)
+	if err != nil {
+		return v1alpha1.ResolvedStack{}, err
+	}
+
+	runMixins, err := readMixins(runImage)
+	if err != nil {
+		return v1alpha1.ResolvedStack{}, err
+	}
+
+	mixins, err := mixins(buildMixins, runMixins)
+
+	return v1alpha1.ResolvedStack{
+		BuildImage: v1alpha1.StackStatusImage{
+			LatestImage: buildIdentifier,
+		},
+		RunImage: v1alpha1.StackStatusImage{
+			LatestImage: runIdentifier,
+		},
+		Mixins:  mixins,
+		UserID:  userId,
+		GroupID: groupId,
+	}, err
+}
+
+func validateStackId(stackId string, buildImage ggcrv1.Image, runImage ggcrv1.Image) error {
+	buildStack, err := imagehelpers.GetStringLabel(buildImage, StackLabel)
+	if err != nil {
+		return err
+	}
+
+	runStack, err := imagehelpers.GetStringLabel(runImage, StackLabel)
+	if err != nil {
+		return err
+	}
+
+	if (buildStack != stackId) || (runStack != stackId) {
+		return errors.Errorf("invalid stack images. expected stack: %s, build image stack: %s, run image stack: %s", stackId, buildStack, runStack)
+	}
+
+	return nil
+}
+
+func readMixins(image ggcrv1.Image) ([]string, error) {
+	var mixins []string
+	hasLabel, err := imagehelpers.HasLabel(image, MixinsLabel)
+	if !hasLabel || err != nil {
+		return mixins, err
+	}
+
+	err = imagehelpers.GetLabel(image, MixinsLabel, &mixins)
+	return mixins, err
+}
+
+func mixins(build, run []string) ([]string, error) {
+	invalid, buildStage, buildCommon := classifyMixins(build)
```

Comment:
`invalid` swaps places when classifying for build and run. That is a bit confusing. Can we pass in a flag to classifyMixins so that it spits out either build or run but not both?

Reply:
Good catch. I like that idea. I tried passing in the mixin prefixes. 

Let me know what you think.